neuraxle/runners.py

```diff
@@ -0,0 +1,277 @@
+            steps_as_tuple=self.steps_as_tuple,
+            data_inputs=data_inputs
+        )
+
+        steps_left_to_do = self.steps_as_tuple[index_starting_step:]
+        data_inputs = self.transform_steps_left_to_do(data_inputs, steps_left_to_do)
+
+        return data_inputs
+
+    def transform_steps_left_to_do(self, data_inputs, steps_left_to_do):
+        """
+        After loading the last checkpoint, transform each pipeline steps left to do
+        :param data_inputs: the data input to transform on
+        :param steps_left_to_do: steps left to do
+        :return: the pipeline itself
+        """
+        for step_name, step in steps_left_to_do:
+            if isinstance(step, TruncableSteps) or isinstance(step, MetaStepsMixin):
+                data_inputs = self.transform_steps_left_to_do_recursive(step, data_inputs)
+            else:
+                data_inputs = step.transform(data_inputs)
+
+        return data_inputs
+
+    def transform_steps_left_to_do_recursive(self, step, data_inputs):
+        """
+        After loading the last checkpoint, recursively transform a step that contains steps
+        :param data_inputs: the data input to transform on
+        :param step: pipeline step
+        :return: the pipeline itself
+        """
+        checkpoint, index = self.find_starting_step_and_data_inputs(
+            steps_as_tuple=step.steps_as_tuple,
+            data_inputs=data_inputs
+        )
+
+        checkpoint_data_input, checkpoint_expected_outputs = checkpoint
+        if checkpoint_data_input is not None:
+            steps_left_to_do_recursive = step.steps_as_tuple[index:]
+            return self.transform_steps_left_to_do(checkpoint_data_input, steps_left_to_do_recursive)
+
+        return step.transform(data_inputs)
+
+    def find_starting_step_and_data_inputs(self, steps_as_tuple, data_inputs, expected_outputs=None):
+        """
+        Find the starting step index, and its corresponding data inputs using the checkpoint steps.
+        If the checkpoint is inside another step, the starting step will be step that contains the checkpoint
+        :param steps_as_tuple: steps
+        :param expected_outputs: expected outputs to fit on
+        :param data_inputs: the data input to fit on
+        :return: tuple(tuple(data_inputs, expected_outputs), step_index) tuple for the starting step data inputs-outputs,
+         and the starting step index
+        """
+        new_data_inputs = data_inputs
+        new_expected_outputs = expected_outputs
+        new_starting_step_index = 0
+        checkpoint_new_data_inputs = None
+        checkpoint_new_expected_outputs = None
+
+        for index, (step_name, step) in enumerate(reversed(steps_as_tuple)):
+            if isinstance(step, BaseCheckpointStep):
+                checkpoint_new_data_inputs, checkpoint_new_expected_outputs = self.load_checkpoint(step)
+
+            if isinstance(step, TruncableSteps) or isinstance(step, MetaStepsMixin):
```

Comment:
TDA again. It seems to me that this code could be simplified so much by respecting the TDA. 

Suggested fix: 
```
if isinstance(step, Pipeline) and step.is_resumable():
    steps_left_to_do = step.get_subset_for_resuming_pipeline()
```

Note: `is_resumable` is suggested but we could come up with a better name. Also, the `get_subset_for_resuming_pipeline()` would return a slice of the Truncable pipeline step such that it is already reduced as `steps_left_to_do`. This would perhaps greatly simplify code. 

If the step is not resumable, then it means we need to continue and test the other ones in reverse.

Reply:
This algo would at least work for transforming: its fine if steps recursively trim themselves like that, the recursion would be within `get_subset_for_resuming_pipeline(...)`. I dont know for fitting because when we fit we need to override the self with the returned object fitted.

Reply:
Here is what I plan on doing : 

Move the following methods inside resumable steps classes(MetaStepsMixin, TruncableSteps) :  

- **fit_transform_steps_left_to_do_recursive** (this would eliminate **fit_transform_steps_left_to_do** method as well)

- **transform_steps_left_to_do_recursive** (this would eliminate **transform_steps_left_to_do** method as well)

Create a load_checkpoint method for resumable steps classes(MetaStepsMixin, TruncableSteps) that loads the checkpoints recursively.

Extract the method **find_starting_step_and_data_inputs** inside a pure fonction so that all these classes can use it when needed : MetaStepsMixin, TruncableSteps, CheckpointPipelineRunner

Reply:
Can you do this in the pipeline object or a similar object that can use self as in `self.steps_as_tuple`, perhaps in the class Pipeline? After thought, I don't think it goes into TruncableSteps as TruncableSteps isn't a forward sequence that can be resumed as a pipeline, it's mostly a list of steps. See: 

```
    def set_path(self, checkpoint_cache_path: str): 
        # This would be called by the pipeline runner directly after `.set_steps` or as a `.setup()` thing or something, just once at the beginning. 
        self.checkpoint_cache_path: str = checkpoint_cache_path

    def transform(self, data_inputs):
        truncated_self = self.resume_transform_entry_point(data_inputs)

        for step_name, step in truncated_self.steps:
            data_inputs = step.transform(data_inputs)
        processed_outputs = data_inputs
        return processed_outputs

    def resume_transform_entry_point(self, data_inputs):
        steps = copy(self.steps_as_tuple)
        for step_name, step in reversed(steps_as_tuple):

            if hasattr(step, 'should_resume') and step.should_resume(data_inputs):
                truncated_self = self[step_name:]
                if hasattr(truncated, 'trim_for_resume'):
                    truncated_self = truncated_self.resume_transform_entry_point(data_inputs)
                return truncated_self

    def should_resume(data_inputs): 
        return True  # TODO: check if we have a checkpoint that has effectively saved its things to disk, etc. Using `self.checkpoint_cache_path` or something like that perhaps. 
```

Reply:
> Extract the method find_starting_step_and_data_inputs inside a pure fonction so that all these classes can use it when needed : MetaStepsMixin, TruncableSteps, CheckpointPipelineRunner

I think this would be a base class method not a pure function.

Reply:
> `set_checkpoint_name`

This too, (base class method, not pure function)

Reply:
@guillaume-chevalier there is no state needed in these methods so we don't need a base class method I think

Reply:
See the `truncated_self = self[step_name:]` in my code above: it somehow feels good to be able to truncate like that. That is the `find_starting_step_and_data_inputs` method you talk about (in my case named like `resume_transform_entry_point`), it seems to use `self`. For the `set_checkpoint_name` and for the `steps_as_tuple`, I find it hard to see a setter method not set anything in `self`. I'll wait for your next changes as it seems you have started coding. I would avoid mixing functional to OOP for uniformity.